runtime/near-vm-runner/src/cache.rs

```diff
@@ -1,37 +1,137 @@
+        Err(_e) => near_primitives::hash::hash(code),
+    };
+    let key = ContractCacheKey::Version1 {
+        code_hash: hash,
+        vm_config_non_crypto_hash: config.non_crypto_hash(),
+        vm_kind: vm_kind,
+    };
+    near_primitives::hash::hash(&key.try_to_vec().unwrap())
+}
+
+fn cache_error(error: VMError, key: &CryptoHash, cache: &dyn CompiledContractCache) -> VMError {
+    let record = CacheRecord::Error(error.clone());
+    if cache.put(&(key.0).0, &record.try_to_vec().unwrap()).is_err() {
+        // That's fine, just cannot cache compilation error.
+        panic!("Cannot cache an error");
+    }
+    error
+}
+
+fn compile_and_serialize_wasmer(
+    wasm_code: &[u8],
+    config: &VMConfig,
+    key: &CryptoHash,
+    cache: &dyn CompiledContractCache,
+) -> Result<wasmer_runtime::Module, VMError> {
+    let module = compile_module(wasm_code, config).map_err(|e| cache_error(e, &key, cache))?;
+    let artifact =
+        module.cache().map_err(|_e| VMError::CacheError(SerializationError { hash: (key.0).0 }))?;
+    let code = artifact
+        .serialize()
+        .map_err(|_e| VMError::CacheError(SerializationError { hash: (key.0).0 }))?;
+    // If errors comes from serialization we shall not cache it.
+    let serialized = CacheRecord::Code(code)
+        .try_to_vec()
+        .map_err(|_e| VMError::CacheError(SerializationError { hash: (key.0).0 }))?;
+    cache.put(key.as_ref(), &serialized).map_err(|_e| VMError::CacheError(WriteError))?;
+    Ok(module)
+}
+
+fn deserialize_wasmer(serialized: &[u8]) -> Result<wasmer_runtime::Module, VMError> {
+    let record = CacheRecord::try_from_slice(serialized)
+        .map_err(|_e| VMError::CacheError(DeserializationError))?;
+    let code = match record {
+        CacheRecord::Error(err) => return Err(err),
+        CacheRecord::Code(code) => code,
+    };
+    let artifact = Artifact::deserialize(code.as_slice())
+        .map_err(|_e| VMError::CacheError(DeserializationError))?;
+    unsafe {
+        let compiler = compiler_for_backend(Backend::Singlepass).unwrap();
+        load_cache_with(artifact, compiler.as_ref())
+            .map_err(|_e| VMError::CacheError(DeserializationError))
+    }
+}
+
+pub(crate) fn compile_module_cached_wasmer(
+    wasm_code_hash: &[u8],
+    wasm_code: &[u8],
+    config: &VMConfig,
+    cache: Option<&dyn CompiledContractCache>,
+) -> Result<wasmer_runtime::Module, VMError> {
+    /* Consider adding `|| cfg!(feature = "no_cache")` */
+    if cache.is_none() {
+        return compile_module(wasm_code, config);
+    }
+    let key = get_key(wasm_code_hash, wasm_code, VMKind::Wasmer, config);
+    let cache = cache.unwrap();
+    match cache.get(&(key.0).0) {
+        Ok(serialized) => match serialized {
+            Some(serialized) => match deserialize_wasmer(serialized.as_slice()) {
+                Ok(module) => Ok(module),
+                Err(e) => {
+                    panic!("Cannot deserialize cached contract: {:?}", e);
```

Comment:
Sorry. I might have explained incorrectly. I didn't mean to panic here, I mean the node will panic eventually if you return something like `VMError::CacheError(DeserializationError)` that will be turned into a `StorageError::InconsitentStateError` that will panic the node in Nightshare RuntimeAdapter.